transaction-kafka-samples/README.adoc

```diff
@@ -0,0 +1,47 @@
+== Spring Cloud Stream Sample Transactional Application
+
+This sample is showing how you can get a transactional producer using Kafka.
+
+The flow is as follows
+
+1. POST request with /<name> to :9997
+2. transaction-http-source put on kafka topic person-event
+3. transaction-spring-data-processor
+a. listen to topic person-event
+b. create new person entity and save to maria db
+c. create a new person saved event and put on topic person-command
+d. all this as a single transaction
+4. transaction-logger-sink litsen to person-command and log out
+
+=== How to start sample
+Start kafka, zookeeper and mariaDb with `docker-compuse up -d` in this folder. 
+Start each of the individual submodules in this project. Either in your IDE or by running `mvn spring:boot run` in each subfolder.
+Send a POST request to localhost:9997
+
+=== How to verify success and failure
+Do a post request to localhost:9997 with a name field. This can be done with the following code using https://httpie.org[httpie]
+```
+http :9997 name=TonyStark
+```
+In the log for the sink application you will see that there is a PersonCreated event, the person is stored in the database.
+Repeat the same command again and observe in the same log that there is not another PersonCreated event with the same name.
+
+Look in the processor log and observe that the message is not removed from the queue since its transaction fails.
+
+NB! Note that in a real world application you would have to handle unique constraint errors like this in the producers. Transaction failures like this are there to catch errors that you do not expect will happen, like network errors or similar scenarios. 
+
+=== What properties makes this request transactional
+ - Add @Transactional to method you want to be transactional and @EnableTransactional on the class
+ - Configure the Producing application. 
+```
+spring.kafka.binder.transaction:
+  transaction-id-prefix: person-
+  producer:
+    configuration:
+      retries: 1
+      acks: all
+```
+
+Transaction-id-prefix is required to start a transaction, The binder also requires setting retires to anything but 0 and set acks to all in order to enable transactions.
+
+Note that consumers of the resulting topic should set issolation level to READ_COMMITED. This will not read messages that are on the topic but that is not commited.
```

Comment:
General comment on running the samples: The first time I sent the `http` command with TonyStark,everything looks fine. When I sent it again, nothing comes on logging sink as expected, but the processor keeps re-trying infinitely. Any new `http` command with new values don't work as well (although I see them on the http source logs). Could you take a look at them and verify if that is the same behavior that you are seeing?

Reply:
Either the application needs to keep some state, somehow, on how many times it's seen (and rejected) the same offset or a custom `AfterRollbackProcessor` has to be wired into the listener container to do the same - keep some state on how may times it's seen the record for the same partition/offset.

The `DefaultAfterRollbackProcessor` maintains no state so the retries are infinite.

We should probably add a stateful processor to Spring Kafka and add some configuration option to the binder to wire it in. https://github.com/spring-projects/spring-kafka/issues/763

In the meantime, I think the sample could add some simple logic, perhaps with an `AtomicBoolean` field to alternate failure/not failure.

Reply:
Thanks, @garyrussell. 
@bjartek Do you think you can address Gary's suggestion in the PR. As soon as that change is in, this is ready for merging.

Reply:
Will try to fix it next week! Thanks for the pointers @garyrussell!